File: BAEKJOON/9019.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <string>
#include <tuple>

using namespace std;

int D(int n) {
    return (2 * n) % 10000;
}

int S(int n) {
    return (n == 0) ? 9999 : n - 1; //if n==0 이면 9999 else : n-1
}

int L(int n) {
    return (n % 1000) * 10 + n / 1000;
}

int R(int n) {
    return (n % 10) * 1000 + n / 10;
}

string bfs(int A, int B) {
    queue<pair<int, string>> q;
    vector<bool> visited(10000, false);
    
    q.push({A, ""});
    visited[A] = true;
    
    while (!q.empty()) {
        int current;
        string operations;
        tie(current, operations) = q.front();
        q.pop();
        
        if (current == B) {
            return operations;
        }
        
        // D operation
        int next_num = D(current);
        if (!visited[next_num]) {
            visited[next_num] = true;
            q.push({next_num, operations + "D"});
        }
        
        // S operation
        next_num = S(current);
        if (!visited[next_num]) {
            visited[next_num] = true;
            q.push({next_num, operations + "S"});
        }
        
        // L operation
        next_num = L(current);
        if (!visited[next_num]) {
            visited[next_num] = true;
            q.push({next_num, operations + "L"});
        }
        
        // R operation
        next_num = R(current);
        if (!visited[next_num]) {
            visited[next_num] = true;
            q.push({next_num, operations + "R"});
        }
    }
    
    return "";
}

int main() {
    int T;
    cin >> T;

    while (T--) {
        int A, B;
        cin >> A >> B;
        cout << bfs(A, B) << endl;
    }

    return 0;
}

File: BAEKJOON/9019.py
from collections import deque

def D(n):
    return (2 * n) % 10000

def S(n):
    return 9999 if n == 0 else n - 1

def L(n):
    return (n % 1000) * 10 + n // 1000

def R(n):
    return (n % 10) * 1000 + n // 10

def bfs(A, B):
    queue = deque([(A, '')])
    visited = [False] * 10000
    visited[A] = True
    
    while queue:
        current, operations = queue.popleft()
        
        if current == B:
            return operations
        
        # Apply the D operation
        next_num = D(current)
        if not visited[next_num]:
            visited[next_num] = True
            queue.append((next_num, operations + 'D'))
        
        # Apply the S operation
        next_num = S(current)
        if not visited[next_num]:
            visited[next_num] = True
            queue.append((next_num, operations + 'S'))
        
        # Apply the L operation
        next_num = L(current)
        if not visited[next_num]:
            visited[next_num] = True
            queue.append((next_num, operations + 'L'))
        
        # Apply the R operation
        next_num = R(current)
        if not visited[next_num]:
            visited[next_num] = True
            queue.append((next_num, operations + 'R'))
    
    return ""

T = int(input().strip())

for _ in range(T):
    A, B = map(int, input().strip().split())
    result = bfs(A, B)
    print(result)
